### ISSUE TITLE
Audit Web Environment Variables DX and Current Functionality

### ISSUE THREAD

**Issue by thedavidprice** [redwoodjs/graphql#4503]
Related to work on #4334 and #4461

cc @dac09 

## Discussion Topics
- What is the DX we want for Web env var?
- Should we consider .env.default for build (i.e. production deploy)

## Deployment test
`console.log(process.env.X)` where X = ...:
1. env var that isn’t defined anywhere
2. env var that’s defined in hosting provider dashboard
3. env var that’s defined in toml
4. env var that has a prefix
5. evn var that’s in env.default

## Goals:
- [ ] Define the optimal DX working with Web env vars across local and production contexts
- [ ] Audit 1) current behavior of env var handling for deployment and 2) documentation for using Web env vars
- [ ] Add CI (unit or functional?) to prevent regression

# Deployed Test
> NOTE: 
> As of now the only Env defined in the Netlify Dashboard is `DOT_ENV`
> 
> Need to add `TOML_ARRAY` and `REDWOOD_ENV_PREFIX` to Netlify deploy dashboard environment

## Netlify Site
https://hardcore-mestorf-245ef2.netlify.app
## GitHub Repo
https://github.com/redwoodjs/env-var-test

Home page:
https://github.com/redwoodjs/env-var-test/blob/main/web/src/pages/HomePage/HomePage.tsx

# Local Test
## `yarn rw dev`
<img width="762" alt="Screen Shot 2022-02-16 at 11 55 28 AM" src="https://user-images.githubusercontent.com/2951/154350361-9d7d04e9-2ea9-4992-a896-1f5769fb451d.png">

## `yarn rw serve` (after building)
<img width="970" alt="Screen Shot 2022-02-16 at 11 56 41 AM" src="https://user-images.githubusercontent.com/2951/154350333-46766ec0-4b8f-48e2-b680-64b5cbbd6fa3.png">

**Comment by thedavidprice** [comment]
@dac09 So far, it's all working in production according to docs.

However, I am confused as to why locally I can't access value when using `.env.defaults` — it was working, then I added a `.env` and then I could never get it working again even after git clean and deleting .env. Needs diagnosing.

I suspect once I add `TOML_ARRAY` and `REDWOOD_ENV_PREFIX` to Netlify dashboard they'll be working. But wanted to start here.

**Comment by Tobbe** [comment]
![image](https://user-images.githubusercontent.com/30793/154417097-8c1c3f11-b408-4897-8544-69fce53b7293.png)

This is wrong, right? `DOT_ENV` should not exist, because it's neither in your toml nor prefixed with `REDWOOD_ENV_`

![image](https://user-images.githubusercontent.com/30793/154417392-2e2e903e-fd9d-4e62-acf4-d46e8200bdee.png)

`DOT_ENV` is neither 1 nor 2, so should not be available to the web side (but it should be available to the api side)

That's how I'm interpreting the docs at least

**Comment by dac09** [comment]
Precisely the problem I was pointing out @Tobbe  - there's inconsistencies between how we're saying it should work, and how it actually does.

**Comment by thedavidprice** [comment]
...in production...

The docs have always referred to production settings. Maybe not clearly enough. 

Agree that the discrepancy is confusing.

**Comment by Tobbe** [comment]
Yes, in production. `yarn rw serve` is production, no? That's why I made a point of including your red "yarn rw serve" text when I used part of your screenshots.

For `yarn rw dev` I think your screenshot is correct.

**Comment by virtuoushub** [comment]
To add some color here _(I think this is the right place, if not apologies and will open a separate ticket)_, `NODE_ENV` being set to `production` can cause some undesirable behavior when deploying Storybook, specifically [interaction testing](https://storybook.js.org/blog/interaction-testing-with-storybook/) (see: https://github.com/redwoodjs/example-storybook/pull/32).

If I try and [manually override this env var](https://github.com/redwoodjs/example-storybook/pull/32/commits/816ca2d33361decb3d5824c0f7a07843438ec676), bad things happen. Code such listed below  makes it so Storybook builds only sometimes work in non production environments.

- https://github.com/redwoodjs/redwood/blob/main/packages/core/config/webpack.common.js#L133-L139
- deltas in https://github.com/redwoodjs/redwood/blob/main/packages/core/config/webpack.development.js / https://github.com/redwoodjs/redwood/blob/main/packages/core/config/webpack.production.js
- https://github.com/redwoodjs/redwood/blob/main/packages/testing/config/storybook/main.js#L142-L145
